node/worker/worker.go

```diff
@@ -262,10 +262,15 @@ func (w *Worker) GetNewEpoch() *big.Int {
 	parent := w.chain.CurrentBlock()
 	epoch := new(big.Int).Set(parent.Header().Epoch())
 
-	// TODO: Don't depend on sharding state for epoch change.
-	if len(parent.Header().ShardState()) > 0 && parent.NumberU64() != 0 {
-		// ... except if parent has a resharding assignment it increases by 1.
-		epoch = epoch.Add(epoch, common.Big1)
+	shardState, err := parent.Header().GetShardState()
+	if err == nil && shardState.Epoch != nil {
+		// For shard state of staking epochs, the shard state will have an epoch and it will decide the next epoch for following blocks
+		epoch = new(big.Int).Set(shardState.Epoch)
+	} else {
```

Comment:
when does the "else" case happen, should we require shardState.Epoch always not empty?

Reply:
For pre-staking case, the state.Epoch is nil. After staking, it's not nil.

Reply:
In this case, it's better to use IsStakeEpoch instead of checking it's nil, which just depending on the proposer

Reply:
Already added the check in local. Won't bother triggering another jenkins. Will add in following PR.